src/wallet/coin_selection.rs

```diff
@@ -116,6 +87,9 @@ use std::convert::TryInto;
 pub type DefaultCoinSelectionAlgorithm = BranchAndBoundCoinSelection;
 #[cfg(test)]
 pub type DefaultCoinSelectionAlgorithm = LargestFirstCoinSelection; // make the tests more predictable
+/// Algorithm to use in case of an error in the chosen coin selection algorithm
+// If everything fails, simple random selection should work
+pub type FallbackCoinSelectionAlgorithm = SingleRandomDrawCoinSelection;
```

Comment:
First, I agree with Murch that it would have been better to have this split in a different commit - but I understand it's not easy to see when to squash and when not to squash :) If you can, review would be much easier with two commits!

Second, this is a bit suboptimal: not every CS algorithm has a fallback algorithm! Instead, me and @afilini came up with this idea of having a `FallbackWrapper` algorithm struct (you should be able to copy-paste this code and it should already work):
```diff
diff --git a/src/wallet/coin_selection.rs b/src/wallet/coin_selection.rs
index 49bafa9..ef6139b 100644
--- a/src/wallet/coin_selection.rs
+++ b/src/wallet/coin_selection.rs
@@ -84,12 +84,10 @@ use std::convert::TryInto;
 /// Default coin selection algorithm used by [`TxBuilder`](super::tx_builder::TxBuilder) if not
 /// overridden
 #[cfg(not(test))]
-pub type DefaultCoinSelectionAlgorithm = BranchAndBoundCoinSelection;
+pub type DefaultCoinSelectionAlgorithm =
+    FallbackWrapper<BranchAndBoundCoinSelection, SingleRandomDrawCoinSelection>;
 #[cfg(test)]
 pub type DefaultCoinSelectionAlgorithm = LargestFirstCoinSelection; // make the tests more predictable
-/// Algorithm to use in case of an error in the chosen coin selection algorithm
-// If everything fails, simple random selection should work
-pub type FallbackCoinSelectionAlgorithm = SingleRandomDrawCoinSelection;
 
 // Base weight of a Txin, not counting the weight needed for satisfying it.
 // prev_txid (32 bytes) + prev_vout (4 bytes) + sequence (4 bytes)
@@ -250,25 +248,13 @@ pub fn get_selection<D: Database, Cs: CoinSelectionAlgorithm<D>>(
         } else {
             // from now on, target_amount can only be positive
             let target_amount = (target_amount_i64 - required_effective_value) as u64;
-            let (mut selected_utxos, selected_fees) = match algorithm.coin_select(
+            let (mut selected_utxos, selected_fees) = algorithm.coin_select(
                 database,
                 optional_utxos.clone(),
                 fee_rate,
                 target_amount,
                 optional_effective_value,
-            ) {
-                Ok(selection) => selection,
-                Err(_err) => {
-                    let fallback_algorithm = FallbackCoinSelectionAlgorithm {};
-                    fallback_algorithm.coin_select(
-                        database,
-                        optional_utxos,
-                        fee_rate,
-                        target_amount,
-                        optional_effective_value,
-                    )?
-                }
-            };
+            )?;
             let selected_effective_value: i64 =
                 selected_utxos.iter().map(|og| og.effective_value).sum();
             selected_utxos.append(&mut required_utxos);
             @@ -344,6 +330,59 @@ pub trait CoinSelectionAlgorithm<D: Database>: std::fmt::Debug {
     ) -> Result<(Vec<OutputGroup>, u64), Error>;
 }
 
+#[derive(Debug)]
+pub struct FallbackWrapper<M, F> {
+    main: M,
+    fallback: F,
+}
+
+impl<M, F> Default for FallbackWrapper<M, F>
+where
+    M: Default,
+    F: Default,
+{
+    fn default() -> Self {
+        FallbackWrapper {
+            main: M::default(),
+            fallback: F::default(),
+        }
+    }
+}
+
+impl<M, F, D> CoinSelectionAlgorithm<D> for FallbackWrapper<M, F>
+where
+    M: CoinSelectionAlgorithm<D>,
+    F: CoinSelectionAlgorithm<D>,
+    D: Database,
+{
+    fn coin_select(
+        &self,
+        database: &D,
+        optional_utxos: Vec<OutputGroup>,
+        fee_rate: FeeRate,
+        target_amount: u64,
+        available_value: i64,
+    ) -> Result<(Vec<OutputGroup>, u64), Error> {
+        self.main
+            .coin_select(
+                database,
+                optional_utxos.clone(),
+                fee_rate,
+                target_amount,
+                available_value,
+                )
+            })
+    }
+}
+
 /// Simple and dumb coin selection
 ///
 /// This coin selection algorithm sorts the available UTXOs by value and then picks them starting
```

This is basically introducing a `FallbackWrapper`, which you can build around your favorite CS algorithms. We implemented CoinSelectionAlgorithm on the wrapper, so that you just need to call `coin_select`, and we automatically take care of calling the fallback in case of errors.

Reply:
I just noticed that calling `BranchAndBounCoinSelection::default()` is suboptimal (it's using a default `size_of_change`)... I'm working on finding a solution...

Reply:
I may be missing something, but shouldn't be a call to `self.callback.coin_select` inside the impl of `coin_select` for `FallbackWrapper`?

Reply:
Sorry, I probably copied it wrong 🤦🏻‍♀️ Yeah, it should be:
```diff
+impl<M, F, D> CoinSelectionAlgorithm<D> for FallbackWrapper<M, F>
+where
+    M: CoinSelectionAlgorithm<D>,
+    F: CoinSelectionAlgorithm<D>,
+    D: Database,
+{
+    fn coin_select(
+        &self,
+        database: &D,
+        optional_utxos: Vec<OutputGroup>,
+        fee_rate: FeeRate,
+        target_amount: u64,
+        available_value: i64,
+    ) -> Result<(Vec<OutputGroup>, u64), Error> {
+        self.main
+            .coin_select(
+                database,
+                optional_utxos.clone(),
+                fee_rate,
+                target_amount,
+                available_value,
+            )
+            .or_else(|_| {
+                self.fallback.coin_select(
+                    database,
+                    optional_utxos.clone(),
+                    fee_rate,
+                    target_amount,
+                    available_value,
+                )
+            })
+    }
+}
+
```

Reply:
> I just noticed that calling `BranchAndBounCoinSelection::default()` is suboptimal (it's using a default `size_of_change`)... I'm working on finding a solution...

But the user can still pass `algorithm` as:
```rust
DefaultCoinSelectionAlgorithm {
   main: BranchAndBoundCoinCoinSelection::new(<user-set-size-of-change>),
   fallback: SingleRandomDrawCoinSelection::default()
}
```
This default is not giving up the flexibility above.

Reply:
I'll continue working with this in another PR.